Exit failure on CommandError or other wrapped error

ekimekim opened the pull request:
Upstreams https://github.com/ekimekim/yaargh/pull/12
Fixes https://github.com/neithere/argh/issues/118

This is a breaking change.

The old behavior was to return successfully from `dispatch()` if a CommandError or wrapped error was caught, not explicitly to exit with status 0.

In the vast majority of cases, users are writing something like:
```python
if __name__ == '__main__':
    argh.dispatch_command(mycommand)
```
and so when `dispatch()` returns, the interpreter almost always exits afterwards due to execution completing.

In order to control the exit code, we had to change this behaviour to explicitly `sys.exit()` on error. This might break users who are doing something after dispatch returns, but it's worth noting that `dispatch()` could already raise SystemExit if parsing failed or `--help` was given (this happens inside argparse itself).

Of course, the other breaking aspect of this change is that commands that used to appear to succeed will now report failure, but the cases where this is undesired are greatly outnumbered by the cases where the current behaviour is a latent bug.

Finally, we include the option to customize the exit status via a new keyword arg `code` to `CommandError`. This shouldn't be able to collide with any existing usage, because previously all keyword args to `CommandError` were rejected.
The name `code` for the keyword arg matches the same arg in `SystemExit`.

#### Upstreaming notes

The cherry-pick onto the latest version of `argh` turned out to be pretty clean, though I had to merge the `CHANGES` manually. I also updated the tutorial docs which I failed to do in my original PR.

neithere commented:
Awesome, thank you!

This will go to Argh v.0.29, so I've just merged the main PR that forms v.0.28 (argh#157). Unfortunately this has created some conflicts. I checked them in advance and they all seem to be fairly trivial, so I hope it's not a big problem.

ekimekim commented:
Updated. I got very confused by the testing changes for a while before re-discovering how I solved them last time, once I caught up to what my previous self was doing it was smooth sailing.
I updated the `CHANGES` to be for version `0.29.0`, and I also removed some python2-compat code around keyword-only arguments.

neithere commented:
Thank you! Looks good. I've only fixed the styling errors and removed warnings which are invalidated by this fix.

Merging and hoping that squashing won't erase the info about the author of original commits...
